theano/tensor/c_code/alt_blas_template.c

```diff
@@ -257,3 +259,33 @@ void %(precision)sgemv_(
     Py_XDECREF(matrixX);
     Py_XDECREF(matrixA);
 }
+
+/* dot */
+%(float_type)s %(precision)sdot_(
+    const int* N,
+    %(float_type)s *SX,
+    const int *INCX,
+    %(float_type)s *SY,
+    const int *INCY
+) {
+    if (*N < 0 || *INCX < 0 || *INCY < 0)
+        alt_fatal_error("The integer arguments passed to %(precision)sdot_ must all be at least 0.");
+    // Create vector_x with shape (1, N)
+    PyObject* vector_x = alt_op_%(float_type)s(0, SX, 1, *N, *INCX, 0);
+    // Create vector_y with shape (N, 1)
+    PyObject* vector_y = alt_op_%(float_type)s(1, SY, 1, *N, *INCY, 0);
+
+    if (vector_x == NULL || vector_y == NULL)
+        alt_fatal_error("NumPy %(precision)sdot_: unables to wrap x and y arrays.");
+
+    // Make matrix product: (1, N) * (N, 1) => (1, 1)
+    PyObject* dot_product = PyArray_MatrixProduct(vector_x, vector_y);
+    if (dot_product == NULL)
+        alt_fatal_error("NumPy %(precision)sdot_: unables to compute dot.");
+    // Get result.
+    %(float_type)s result = *(%(float_type)s*)PyArray_DATA((PyArrayObject*)dot_product);
+    Py_XDECREF(dot_product);
```

Comment:
This will free the prt result you keep. I think you need to use MatrixProduct2() but allocate the output with malloc first, then wrap it in PyArray and pass it to MatrixProduct2

Reply:
I don't keep the pointer, I dereference and copy value instead (`%(float_type)s result = *(` `...`), as we expect just 1 scalar value.